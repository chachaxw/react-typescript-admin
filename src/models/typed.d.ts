// Global declare
declare global {
  interface Window {
    ga: any;
    WxLogin: any;
  }
}

// Global auth interface
export interface Auth {
  isAuthenticated: boolean;
  permissions: string[];
}
