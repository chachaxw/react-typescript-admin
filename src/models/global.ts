// Global declare
declare global {
  interface Window { ga: any; }
}

// Global auth interface
export interface Auth {
  isAuthenticated: boolean;
  permissions: string[];
}

// Global userInfo interface
export interface UserInfo {
  name: string;
  mobile: string;
  avatarUrl: string;
  nickname: string;
}

// Global state interface
export interface GlobalState {
  auth: Auth;
  collapsed: boolean;
  notices: any[];
  userInfo: UserInfo;
}

export default {
  namespace: 'global', // model的命名空间

  state: { // 应用的状态数据
    auth: {
      isAuthenticated: true,
      permissions: [],
    },
    userInfo: {
      name: '',
      mobile: '',
      avatarUrl: '',
      nickname: '',
    },
    collapsed: false,
    notices: [],
    hasMore: false,
  },

  effects: { // 异步请求处理和业务逻辑操作

  },

  reducers: { // Redux reducers
    changeLayoutCollapsed(state: GlobalState, { payload }: any) {
      return { ...state, collapsed: payload };
    },

    loadUserInfo(state: GlobalState, { payload }: any) {
      return { ...state, userInfo: payload };
    },

    loadNotices(state: GlobalState, { payload, hasMore }: any) {
      return { ...state, notices: payload, hasMore };
    },
  },

  subscriptions: { // 用于订阅一个数据源, 然后根据需要 dispatch 相应的 action
    setup({ dispatch, history }: any) {
      // TODO: Need to add auth and permissions
      return history.listen(({ pathname, search }: Location) => {
        if (typeof window.ga !== 'undefined') {
          // TODO: Google Analytics config, 埋点配置
          window.ga('send', 'PageView', pathname + search);
        }
      });
    },
  },
};
