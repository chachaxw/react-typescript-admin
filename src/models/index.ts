import { DvaInstance, Model } from 'dva';

export { default as Global } from './global';
export { default as Dashboard } from './dashboard';

// 动态注册 model 并缓存已注册过的 model
const cached = {};

export function registerModel(app: DvaInstance, model: Model) {
  if (!cached[model.namespace]) {
    app.model(model);
    cached[model.namespace] = 1;
  }
}
