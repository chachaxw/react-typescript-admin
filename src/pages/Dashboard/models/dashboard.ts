// Dashboard state
export default {
  namespace: 'dashboard', // model的命名空间

  state: {
    // 应用的状态数据
  },

  effects: {},

  reducers: {
    // Redux reducers
  },

  subscriptions: {
    setup({ dispatch, history }: any) {
      return history.listen(({ pathname }: any) => {
        if (pathname === '/app/dashboard') {
        }
      });
    },
  },
};
