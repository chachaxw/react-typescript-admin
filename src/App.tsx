import './App.css';

import { Layout } from 'antd';
import { connect } from 'dva';
import React, { Component } from 'react';
import DocumentTitle from 'react-document-title';

import { Content, ErrorBoundary, Footer, Header, SideBar } from './components';
import { Auth } from './models/global';
import AppRoutes from './routes';
import routesConfig from './routes/config';
import { getFlatMenuKeys } from './utils/utils';

interface DvaProps {
  auth: Auth;
  location: Location;
  collapsed: boolean;
  onCollapse: (collapsed: boolean) => void;
}

interface InternalProps extends DvaProps {
  app: any;
}

interface InternalState {
  title: string;
}

class App extends Component<InternalProps, InternalState> {
  private menu = routesConfig.app;
  private flatMenuKeys = getFlatMenuKeys(this.menu);

  private constructor(props: InternalProps) {
    super(props);
    this.state = {
      title: '',
    };
  }

  public render() {
    const { title } = this.state;
    const { app, auth, collapsed, location, onCollapse } = this.props;

    return (
      <DocumentTitle title={title}>
        <Layout className="App">
          <SideBar collapsed={collapsed} menu={this.menu} onCollapse={onCollapse}
            flatMenuKeys={this.flatMenuKeys} location={location} />
          <Layout>
            <Header />
            <Content>
              <ErrorBoundary>
                <AppRoutes app={app} auth={auth} />
              </ErrorBoundary>
            </Content>
            <Footer className="Footer" />
          </Layout>
        </Layout>
      </DocumentTitle>
    );
  }
}

const mapStateToProps = ({ global }: any) => {
  return {
    auth: global.auth,
    collapsed: global.collapsed,
  };
};

const mapDispatchToProps = (dispatch: any) => {
  return {
    onCollapse: (collapsed: boolean) => {
      dispatch({
        type: 'global/changeLayoutCollapsed',
        payload: collapsed,
      });
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(App);
