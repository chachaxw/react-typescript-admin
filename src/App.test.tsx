import dva from 'dva';
import { StaticRouter } from 'dva/router';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

import App from './App';
import { Global } from './models';

let app: any;
let wrapper: any;

beforeAll(() => {
  app = dva();
  app.model(Global);
  app.router(() => ({}));
  app.start();
  wrapper = shallow(
      <StaticRouter context={{}}>
          <App store={app._store} location={{pathname: '/app/dashboard'}} />
      </StaticRouter>
  );
});

afterAll(() => {
  wrapper.unmount();
});

describe('App Test', () => {
  it('Capturing Snapshot of APP', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
