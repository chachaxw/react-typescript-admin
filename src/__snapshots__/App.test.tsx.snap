// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Test Capturing Snapshot of APP 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Connect(App)
    location={
      Object {
        "pathname": "/app/dashboard",
      }
    }
    store={
      Object {
        "asyncReducers": Object {},
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "runSaga": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  />
</Router>
`;
